gradle{,_7,_8,_9}{,-unwrapped}: remove callPackage from all-packages.nix

numinit opened the pull request:
Since we now have a more strict nixpkgs-vet, deal with all the Gradle callPackage (including for unwrapped versions) in the package set, instead of continuing to fight against CI. See #421047 for context.
    
Should retain complete compatibility with previous nixpkgs versions, particularly check that there were no messups in the review.

(Related note: did gradleGen and wrapGradle need to be exposed? I've moved them to a let block, that is the only possible incompatibility but nothing was referencing them.)

Testing:

```
nom-build -A gradle_7 -A gradle_7-unwrapped -A gradle_8 -A gradle_8-unwrapped -A gradle_9 -A gradle_9-unwrapped \
    -A gradle -A gradle-unwrapped \
    -A gradle-packages
/nix/store/pfgchfmna0cfrqlxhxb3bmi4k01q6vxb-gradle-7.6.6
/nix/store/7ypr35w5g1c4bp55gn8r1dc5dpyr6x4s-gradle-7.6.6
/nix/store/klws7nsbs5bvxxi79j0m2xbiiw4fxdai-gradle-8.14.3
/nix/store/wzwhbsz5f98pychmffzpflkqp33qls9z-gradle-8.14.3
/nix/store/dj9pn0gxk5wpkcvwslrlxihmq9qz5sxx-gradle-9.0.0
/nix/store/vhjrb76n297mfv4an92yihl4g9kf24dz-gradle-9.0.0
/nix/store/klws7nsbs5bvxxi79j0m2xbiiw4fxdai-gradle-8.14.3
/nix/store/wzwhbsz5f98pychmffzpflkqp33qls9z-gradle-8.14.3
```

Rebuild count should be 0.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [x] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

numinit committed:
gradle{,_7,_8,_9}{,-unwrapped}: remove callPackage from all-packages.nix

Since we now have a more strict nixpkgs-vet, deal with all the Gradle
callPackage (including for unwrapped versions) in the package set,
instead of continuing to fight against CI. See #421047 for context.

Should retain complete compatibility with previous nixpkgs versions,
particularly check that there were no messups in the review.

wolfgangwalther reviewed (commented):
I agree this is *much* cleaner than the one before.

numinit commented on pkgs/by-name/gr/gradle-unwrapped/package.nix:
```diff
@@ -0,0 +1,5 @@
+  gradle_8-unwrapped,
+  ...
+}@args:
+gradle_8-unwrapped.override args
```
Yeah, it does, but we don't need to do it. This was a little extreme.

numinit commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6943,25 +6943,18 @@ with pkgs;
-  gradle_8 = wrapGradle gradle_8-unwrapped null;
-  gradle_9 = wrapGradle gradle_9-unwrapped null;
-  gradle = wrapGradle gradle-unwrapped "gradle-unwrapped";
+  gradle_7 = gradle_7-unwrapped.wrap null;
+  gradle_8 = gradle_8-unwrapped.wrap null;
+  gradle_9 = gradle_9-unwrapped.wrap null;
+  gradle = gradle-unwrapped.wrap "gradle-unwrapped";
```
Yeah, and I realized that we're doubling up on the unwrapped function call to gen' too.

msgilligan commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 436358`
Commit: `d9e90ca7b5532d5bc7ff738f132316a2867e9311`

msgilligan reviewed (commented):
I'm really not (yet) qualified to review these code changes. I'm willing to do some testing though. I'll also rebase my 9.0.0 -> 9.1.0(-rc-1) PR on top of this when the time comes.

I ran `nixpkgs-review pr 436358` on `aarch64-darwin` and `aarc64-linux` and it seemed to be a no-op which seems weird.  (I did a post-result on both of them)

msgilligan commented:
> note the "rebuild: 0" labels for both linux and darwin. … a pure refactor that doesn't change any builds, so that's alright.

This is because the hash of the derivations didn't change and CI figured that out when it generated the `rebuild: 0` labels?

numinit committed:
gradle{,_7,_8,_9}{,-unwrapped}: clean up wrapping

Eliminate duplicate function calls and make the API cleaner.

$ nix-instantiate --eval -A gradle.updateScript
[ <CODE> "--url=https://github.com/gradle/gradle" "--version-regex=^v(\\d+\\.\\d+(?:\\.[1-9]\\d?)?)(\\.0)?$" ]
$ nix-instantiate --eval -A gradle_7.updateScript
error: attribute 'updateScript' in selection path 'gradle_7.updateScript' not found
$ nix-instantiate --eval -A gradle_9.updateScript
error: attribute 'updateScript' in selection path 'gradle_9.updateScript' not found

numinit commented:
Diff against the previous commit that should encapsulate @wolfgangwalther's suggestion:

https://github.com/NixOS/nixpkgs/pull/436358/commits/a8213af876d559ae5cd9a6db0c107bf7bfb3297f

numinit committed:
gradle: unwrapped shouldn't have unwrapped passthru

wolfgangwalther reviewed (approved):
Diff LGTM; no rebuilds.

numinit commented on pkgs/top-level/all-packages.nix:
```diff
@@ -6915,25 +6915,18 @@ with pkgs;
+  gradle_8-unwrapped = gradle-packages.gradle_8;
+  gradle_9-unwrapped = gradle-packages.gradle_9;
+  gradle-unwrapped = gradle-packages.gradle;
 
-  gradle_7 = wrapGradle gradle_7-unwrapped null;
-  gradle_8 = wrapGradle gradle_8-unwrapped null;
-  gradle_9 = wrapGradle gradle_9-unwrapped null;
-  gradle = wrapGradle gradle-unwrapped "gradle-unwrapped";
+  gradle_7 = gradle-packages.gradle_7.wrapped;
+  gradle_8 = gradle-packages.gradle_8.wrapped;
+  gradle_9 = gradle-packages.gradle_9.wrapped;
+  gradle = gradle-packages.gradle.wrapped;
```
Ideally we can make all-packages.nix simpler over time :-)

msgilligan commented:
PR #434086 was rebased on top of this and has also been updated for Gradle 9.1.0-RC-2 (which looks like it will be the final RC)  Note that Gradle 9.1.0 is the first release of Gradle to fully support OpenJDK 25.

msfjarvis commented:
> (Related note: did gradleGen and wrapGradle need to be exposed? I've moved them to a let block, that is the only possible incompatibility but nothing was referencing them.)

FWIW I had been using `gradleGen` in [my NixOS config](https://github.com/msfjarvis/dotfiles/blob/71f59cee4c1a819b2ebe0cd1f6edefe265bef39d/modules/nixos/profiles/desktop/android-dev.nix#L54-L59) to be able to test pre-release versions of Gradle more easily, looks like there were [other people using this as well](https://github.com/search?q=path%3A*.nix%20gradleGen&type=code).

wolfgangwalther commented on pkgs/development/tools/build-managers/gradle/default.nix:
```diff
@@ -1,12 +1,111 @@
+          }
+          // gradle.tests;
+        }
+        // lib.optionalAttrs (this-gradle-unwrapped == gradle-unwrapped) {
```
So it turns out that this check is a problem, because it currently makes Eval fail on staging-next.

That's because `gradle_7` has been marked as insecure, but this line makes it that `gradle_7.meta` can't be evaluated without throwing that insecure error. This should not happen.

This needs to be solved differently.

numinit replied:
Can probably do it similarly to how it was before to break the dependency on the meta. I'll put together another PR. A user mentioned overriding needing to be easier too barring a full wrapper replacement which we're doing in #449037

wolfgangwalther replied:
I have something ready to fix eval in a minute.
